File: docs/RA/RA_0468/source/Teradyne.Archimedes.DAS.Listener/DASMessageListener.cs
// This code is provided 'as is' without any guarantee of its performance, suitability, or reliability.
// The code is not intended to be used for production purposes and has not been optimized. 
// It has also not been implemented following all state-of-the-art practices. 
// The purpose of this code is to demonstrate some capabilities of UltraEdge software. 
// It is intended for demonstration purposes only. The author shall not be held liable 
// for any damages arising from the use of this code. Use of this code is solely at the user's own risk.
// Teradyne company will not be responsible for any damages arising during the use of this code. 
// The code should be treated as confidential and is protected by intellectual property rights.

using Serilog;
using System.Net;
using Teradyne.Archimedes.AMPCommunication.Messages;
using Teradyne.Archimedes.AMPCommunication.Models;
using Teradyne.Archimedes.AMPCommunication.Services;
using static Teradyne.Archimedes.AMPCommunication.AMPEvents;

namespace Teradyne.Archimedes.DAS.Listener
{
    /// <summary>
    /// DASMessageListener is responsible for starting and stopping the DAS HTTP listener,
    /// receiving incoming HTTP requests, and dispatching them to a processor. It manages
    /// event notifications for connection, disconnection, errors, and new messages.
    /// </summary>
    /// <author>Teradyne DIA</author>
    public class DASMessageListener : IDisposable
    {
        /// <summary>
        /// Delegate used for the event related to the DAS
        /// </summary>
        public delegate void DasEventHandler();

        /// <summary>
        /// Delegate when the DAS is facing to an error
        /// </summary>
        /// <param name="exception"></param>
        public delegate void DasErrorHandler(Exception exception);

        /// <summary>
        /// Event when the DAS is connected
        /// </summary>
        public event DasEventHandler Connected = delegate { };

        /// <summary>
        /// Event when the DAS encouterred and error
        /// </summary>
        public event DasErrorHandler Error = delegate { };

        /// <summary>
        /// Event when the DAS is disconnected
        /// </summary>
        public event DasEventHandler Disconnected = delegate { };

        /// <summary>
        /// Event when the DAS received a new message
        /// </summary>
        public event NewMessageDlg NewMessage = delegate { };

        /// <summary>
        /// System for managing the list of messages we want to receive
        /// </summary>
        protected AMPMessageManager _messageList;

        private readonly string _dasurl;
        private HttpListener? _listener;
        private CancellationTokenSource? _ctSource;

        /// <summary>
        /// Gets the DAS listener URL.
        /// </summary>
        public string DASURL => _dasurl;

        /// <summary>
        /// Initializes a new instance of the DASMessageListener class with a DAS URL.
        /// </summary>
        /// <param name="dasurl">The URL to listen to.</param>
        /// <exception cref="ArgumentNullException"></exception>
        public DASMessageListener(string dasurl)
        {
            _messageList = new AMPMessageManager();
            _dasurl = dasurl ?? throw new ArgumentNullException(nameof(dasurl));
        }

        /// <summary>
        /// Starts the HTTP listener and begins processing incoming requests.
        /// </summary>
        /// <returns>True if started successfully, false otherwise.</returns>
        public async Task<bool> Start()
        {
            _ctSource = new CancellationTokenSource();
            CancellationToken cancellationToken = _ctSource.Token;
            _listener = new HttpListener();

            try
            {
                if (string.IsNullOrWhiteSpace(_dasurl))
                {
                    Log.Error("DAS URL is invalid.");
                    return false;
                }

                _listener.Prefixes.Add(DASURL);
                _listener.Start();

                try { Connected(); } catch (Exception ex) { Log.Warning(ex, "Error while invoking Connected event."); }

                AMPMessageProcessor requestProcessor = new(_messageList);
                requestProcessor.NewMessage += RequestProcessor_NewMessage;

                while (!cancellationToken.IsCancellationRequested)
                {
                    try
                    {
                        HttpListenerContext context = await _listener.GetContextAsync();
                        await requestProcessor.ProcessRequest(context);
                    }
                    catch (OperationCanceledException)
                    {
                        // Expected when cancelling token
                    }
                    catch (HttpListenerException hlex)
                    {
                        Log.Error(hlex, "HTTP Listener exception.");
                        try { Error(hlex); } catch (Exception ex) { Log.Warning(ex, "Error while invoking Error event."); }
                        Stop();
                        return false;
                    }
                    catch (Exception ex)
                    {
                        Log.Error(ex, "Unexpected exception in DASListener.");
                        try { Error(ex); } catch (Exception ex2) { Log.Warning(ex2, "Error while invoking Error event."); }
                        Stop();
                        return false;
                    }
                }

                return true;
            }
            finally
            {
                Stop();
            }
        }

        /// <summary>
        /// Stops the HTTP listener and cancels all ongoing tasks.
        /// </summary>
        public void Stop()
        {
            _ctSource?.Cancel();
            _ctSource = null;

            if (_listener != null)
            {
                _listener.Stop();
                _listener.Close();
                _listener = null;
                try { Disconnected(); } catch (Exception ex) { Log.Warning(ex, "Error while invoking Disconnected event."); }
            }
        }

        /// <summary>
        /// Handles new messages received by the message processor and raises the NewMessage event.
        /// </summary>
        /// <param name="cellid">The ID of the cell that sent the message.</param>
        /// <param name="message">The message payload.</param>
        private void RequestProcessor_NewMessage(string cellid, AMPMessageModel message)
        {
            try { NewMessage(cellid, message); } catch (Exception ex) { Log.Warning(ex, "Error while invoking NewMessage event."); }
        }

        /// <summary>
        /// Subscribes to a set of message names, optionally resetting previous subscriptions.
        /// </summary>
        /// <param name="messageNames">List of message names to subscribe to.</param>
        /// <param name="resetFirst">Whether to clear existing subscriptions first.</param>
        public void SubscribeToMessages(List<string> messageNames, bool resetFirst = true)
        {
            try
            {
                if (resetFirst) _messageList.UnsubscribeAll();
                _messageList.Subscribe(messageNames);
            }
            catch (Exception ex)
            {
                Log.Error(ex, "Failed to subscribe to messages.");
            }
        }

        /// <summary>
        /// Disposes the listener and stops it if necessary.
        /// </summary>
        public void Dispose()
        {
            Stop();
        }
    }
}

File: docs/RA/RA_0478/source/DASListener/src/DASMessageListener.cpp
// This code is provided 'as is' without any guarantee of its performance, suitability, or reliability.
// The code is not intended to be used for production purposes and has not been optimized. 
// It has also not been implemented following all state-of-the-art practices. 
// The purpose of this code is to demonstrate some capabilities of UltraEdge software. 
// It is intended for demonstration purposes only. The author shall not be held liable 
// for any damages arising from the use of this code. Use of this code is solely at the user's own risk.
// Teradyne company will not be responsible for any damages arising during the use of this code. 
// The code should be treated as confidential and is protected by intellectual property rights.

#include "DASMessageListener.h"
#include <iostream>
#include <thread>
#include <atomic>
#include <chrono>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <set>

// Note: This implementation uses a simplified HTTP server for demonstration
// In production, consider using libraries like cpp-httplib, Boost.Beast, or Poco

namespace Teradyne {
namespace Archimedes {
namespace DAS {

// Simple HTTP server implementation
class SimpleHttpServer {
public:
    SimpleHttpServer(const std::string& host, int port)
        : host_(host), port_(port), running_(false) {}

    bool start(std::function<void(const std::string&, const std::string&, const std::string&)> handler) {
        if (running_) return false;
        
        handler_ = handler;
        running_ = true;
        
        // In a real implementation, this would use sockets
        // For demonstration purposes, we'll simulate the server
        serverThread_ = std::thread([this]() {
            std::cout << "HTTP Server started on " << host_ << ":" << port_ << std::endl;
            while (running_) {
                std::this_thread::sleep_for(std::chrono::milliseconds(100));
            }
        });
        
        return true;
    }

    void stop() {
        if (!running_) return;
        running_ = false;
        if (serverThread_.joinable()) {
            serverThread_.join();
        }
    }

    bool isRunning() const { return running_; }

    // Simulate receiving a message (for demonstration)
    void simulateMessage(const std::string& method, const std::string& path, const std::string& body) {
        if (handler_ && running_) {
            handler_(method, path, body);
        }
    }

private:
    std::string host_;
    int port_;
    std::atomic<bool> running_;
    std::thread serverThread_;
    std::function<void(const std::string&, const std::string&, const std::string&)> handler_;
};

// Private implementation
class DASMessageListener::Impl {
public:
    Impl(DASMessageListener* parent, const std::string& dasUrl)
        : parent_(parent), dasUrl_(dasUrl), running_(false) {
        parseUrl(dasUrl);
    }

    ~Impl() {
        stop();
    }

    bool start() {
        if (running_) return false;

        server_ = std::make_unique<SimpleHttpServer>(host_, port_);
        
        bool started = server_->start([this](const std::string& method, 
                                              const std::string& path, 
                                              const std::string& body) {
            handleRequest(method, path, body);
        });

        if (started) {
            running_ = true;
            parent_->invokeConnected();
        }

        return started;
    }

    void stop() {
        if (!running_) return;
        
        if (server_) {
            server_->stop();
            server_.reset();
        }
        
        running_ = false;
        parent_->invokeDisconnected();
    }

    bool isRunning() const {
        return running_;
    }

    void subscribeToMessages(const std::vector<std::string>& messageNames, bool resetFirst) {
        if (resetFirst) {
            subscribedMessages_.clear();
        }
        
        for (const auto& msg : messageNames) {
            subscribedMessages_.insert(msg);
        }
    }

private:
    void parseUrl(const std::string& url) {
        // Simple URL parsing (http://host:port/path/)
        std::string temp = url;
        
        // Remove protocol
        size_t pos = temp.find("://");
        if (pos != std::string::npos) {
            temp = temp.substr(pos + 3);
        }
        
        // Extract host and port
        pos = temp.find(":");
        if (pos != std::string::npos) {
            host_ = temp.substr(0, pos);
            temp = temp.substr(pos + 1);
            
            pos = temp.find("/");
            if (pos != std::string::npos) {
                port_ = std::stoi(temp.substr(0, pos));
                basePath_ = temp.substr(pos);
            } else {
                port_ = std::stoi(temp);
            }
        } else {
            pos = temp.find("/");
            if (pos != std::string::npos) {
                host_ = temp.substr(0, pos);
                basePath_ = temp.substr(pos);
            } else {
                host_ = temp;
            }
            port_ = 80;
        }
    }

    void handleRequest(const std::string& method, const std::string& path, const std::string& body) {
        if (method != "POST") return;

        // Parse the URL path to extract cell ID and message name
        // Expected format: /basePath/TEST_CELL/cellid/LOT/lotid/SUBLOT/sublotid/MESSAGE_NAME
        
        std::string cellId = extractCellId(path);
        std::string messageName = extractMessageName(path);
        
        // Check if we're subscribed to this message
        if (!subscribedMessages_.empty() && 
            subscribedMessages_.find(messageName) == subscribedMessages_.end()) {
            return; // Not subscribed to this message
        }

        AMPMessage message;
        message.cellId = cellId;
        message.messageName = messageName;
        message.url = path;
        message.payload = body;
        message.timestamp = getCurrentTimestamp();

        parent_->invokeNewMessage(cellId, message);
    }

    std::string extractCellId(const std::string& path) {
        // Look for TEST_CELL/{cellid} pattern
        size_t pos = path.find("TEST_CELL/");
        if (pos != std::string::npos) {
            std::string temp = path.substr(pos + 10);
            size_t end = temp.find("/");
            if (end != std::string::npos) {
                return temp.substr(0, end);
            }
        }
        return "UNKNOWN";
    }

    std::string extractMessageName(const std::string& path) {
        // The message name is typically the last segment of the URL
        size_t pos = path.rfind('/');
        if (pos != std::string::npos && pos < path.length() - 1) {
            return path.substr(pos + 1);
        }
        return "UNKNOWN";
    }

    std::string getCurrentTimestamp() {
        auto now = std::chrono::system_clock::now();
        auto time = std::chrono::system_clock::to_time_t(now);
        auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
            now.time_since_epoch()) % 1000;

        std::stringstream ss;
        ss << std::put_time(std::gmtime(&time), "%Y-%m-%dT%H:%M:%S");
        ss << '.' << std::setfill('0') << std::setw(3) << ms.count() << 'Z';
        return ss.str();
    }

    DASMessageListener* parent_;
    std::string dasUrl_;
    std::string host_;
    int port_;
    std::string basePath_;
    std::atomic<bool> running_;
    std::unique_ptr<SimpleHttpServer> server_;
    std::set<std::string> subscribedMessages_;
};

// DASMessageListener implementation

DASMessageListener::DASMessageListener(const std::string& dasUrl)
    : dasUrl_(dasUrl),
      pImpl_(std::make_unique<Impl>(this, dasUrl)) {
}

DASMessageListener::~DASMessageListener() {
    stop();
}

bool DASMessageListener::start() {
    return pImpl_->start();
}

void DASMessageListener::stop() {
    pImpl_->stop();
}

void DASMessageListener::subscribeToMessages(const std::vector<std::string>& messageNames, bool resetFirst) {
    pImpl_->subscribeToMessages(messageNames, resetFirst);
}

bool DASMessageListener::isRunning() const {
    return pImpl_->isRunning();
}

void DASMessageListener::invokeConnected() {
    if (connectedCallback_) {
        try {
            connectedCallback_();
        } catch (const std::exception& e) {
            std::cerr << "Error in Connected callback: " << e.what() << std::endl;
        }
    }
}

void DASMessageListener::invokeDisconnected() {
    if (disconnectedCallback_) {
        try {
            disconnectedCallback_();
        } catch (const std::exception& e) {
            std::cerr << "Error in Disconnected callback: " << e.what() << std::endl;
        }
    }
}

void DASMessageListener::invokeError(const std::string& error) {
    if (errorCallback_) {
        try {
            errorCallback_(error);
        } catch (const std::exception& e) {
            std::cerr << "Error in Error callback: " << e.what() << std::endl;
        }
    }
}

void DASMessageListener::invokeNewMessage(const std::string& cellId, const AMPMessage& message) {
    if (newMessageCallback_) {
        try {
            newMessageCallback_(cellId, message);
        } catch (const std::exception& e) {
            std::cerr << "Error in NewMessage callback: " << e.what() << std::endl;
        }
    }
}

} // namespace DAS
} // namespace Archimedes
} // namespace Teradyne
